File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>
#include <cmath>

using namespace std;

struct Disk {
    int timeCost, requestTime, num;
};

struct cmp{
    bool operator() (Disk &a, Disk &b) {
    if (a.timeCost!= b.timeCost) return a.timeCost > b.timeCost; // false 여야 앞으로
    else if (a.requestTime != b.requestTime) return a.requestTime > b.requestTime;
    else return a.num > b.num;
    }
};

struct compare {
    bool operator() (vector<int> a, vector<int> b){
        if (a[0]!=b[0]) return a[0] < b[0]; // 요청 시점 
        else a[1] < b[1]; // 소요 시간
    }
};

int solution(vector<vector<int>> jobs) {
    int answer = 0;
    
    sort(jobs.begin(), jobs.end());
    priority_queue<Disk, vector<Disk>, cmp> pq;
    
    int idx = 0; int time=0;
    while (!pq.empty() || idx < jobs.size()){
        // 1. 대기 큐에 하나도 없다면 대기 큐에 넣기
        if (pq.empty() && idx < jobs.size()) {
            time = jobs[idx][0];
            pq.push({jobs[idx][1], jobs[idx][0], idx++});
        }
        
        // 2. 실행하고 시간 흐름
        auto [timeCost, requestTime, num] = pq.top(); pq.pop();
        time += timeCost;
        answer += (time - requestTime);
        
        // 3. 시간 흐름에 따라 대기 큐에 들어갔어야 하는 거 다 넣음
        while (idx < jobs.size() && jobs[idx][0] <= time){
            pq.push({jobs[idx][1], jobs[idx][0], idx++});
        }
    }
    
    return floor((float)answer / jobs.size());
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
class PriorityQueue {
        constructor(comparator = (a,b) => a-b) {
            this.heap = [];
            this.comparator = comparator;
        }
    
        push(value) {
            this.heap.push(value);
            this.bubbleUp();
        }
        pop() {
            if (this.heap.length===0) return undefined;
            if (this.heap.length===1) return this.heap.pop();
            
            const next = this.heap[0];
            this.heap[0] = this.heap.pop();
            this.bubbleDown();
            
            return next;
        }
        bubbleUp(){
            let index = this.heap.length - 1;
            
            while (index>0) {
                let parent = Math.floor((index-1) / 2);
                if (this.comparator(this.heap[index], this.heap[parent]) < 0){
                    [this.heap[index], this.heap[parent]]= [this.heap[parent], this.heap[index]];
                    index= parent;
                } else break;
            }
        }
        bubbleDown(){
            let index = 0;
            let length = this.heap.length;
            
            while (true){
                let left = (index) * 2 +1;
                let right = (index) * 2 +2;
                let upper= index;
                
                if (left<length && this.comparator(this.heap[left], this.heap[upper]) < 0) upper= left;
                if (right<length && this.comparator(this.heap[right], this.heap[upper]) < 0) upper= right;
                
                if (upper==index) break;
                
                [this.heap[upper], this.heap[index]] = [this.heap[index], this.heap[upper]];
                index= upper;
            }
        }
        length(){
            return this.heap.length;
        }
}

class Task {
    constructor(index, requestTime, runTime){
        this.index= index;
        this.requestTime= requestTime;
        this.runTime= runTime;
    }
}

function solution(jobs) {
    var answer = 0;
       
    const comparator = (a,b) => {
        return a.runTime - b.runTime || a.requestTime - b.requestTime || a.index - b.index;
    }
    
    const pq = new PriorityQueue(comparator);
    jobs.sort((a,b) => a[0] - b[0]);
    
    let jobIdx= 0;
    let time= 0;
    while (jobIdx < jobs.length || pq.length()!=0){
        while (jobIdx < jobs.length && time >= jobs[jobIdx][0]) {
            // 넣을 수 있는 만큼 넣는다
            pq.push(new Task(jobIdx, jobs[jobIdx][0], jobs[jobIdx][1]));
            jobIdx++;
        }
        
        // 다음 작업을 선택한다
        const next = pq.pop();
    
        if (next===undefined) {
            time= jobs[jobIdx][0]; // 요청시각으로 바로 보냄
            continue;
        }
        
        answer += (time+next.runTime)-next.requestTime;
        time += next.runTime;
    }
    
    return Math.floor(answer / jobs.length);
}